File: app/src/main/java/hr/caellian/notestream/data/PlayerService.java
package hr.caellian.notestream.data;

import android.app.Service;
import android.content.Intent;
import android.media.MediaPlayer;
import android.media.session.MediaSession;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.PowerManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyManager;
import android.util.Log;
import android.view.KeyEvent;
import android.widget.Toast;

import java.util.Collections;
import java.util.Timer;
import java.util.TimerTask;

import hr.caellian.notestream.gui.NoteStreamNotification;
import hr.caellian.notestream.NoteStream;
import hr.caellian.notestream.data.playable.Playable;
import hr.caellian.notestream.lib.Constants;
import hr.caellian.notestream.util.RepeatState;

public class PlayerService extends Service implements MediaPlayer.OnCompletionListener{
    // https://github.com/smedic/Android-YouTube-Background-Player/blob/master/app/src/main/java/com/smedic/tubtub/BackgroundAudioService.java
    // http://www.helloandroid.com/tutorials/musicdroid-audio-player-part-ii
    // http://www.tutorialsface.com/2015/08/android-custom-notification-tutorial/

    public static final int DEFAULT_PROGRESS_CHANGE = 5000;

    private static Timer progressTimer = new Timer();

    int queueSize = 0;
    static Playlist pl = Playlist.get(Playlist.TEMPORARY_PREFIX + "currentlyPlayed");
    static MediaPlayer mp = new MediaPlayer();
    MediaSession ms;
    static RepeatState repeatState = RepeatState.NONE;

    boolean playing;

    static PlayerServiceBinder psb = null;

    NoteStreamNotification notification;

    Handler progressHandler;

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        if (repeatState != RepeatState.ONE) {
            if (pl.getCurrentPlayable() == pl.getLastPlayable() && repeatState != RepeatState.ALL) {
                psb.pause();
            } else {
                psb.switchNext();
            }
        }
    }

    private static class ProgressHandler extends Handler {
        MediaPlayer mp;

        public ProgressHandler(MediaPlayer mediaPlayer) {
            super();
            mp = mediaPlayer;
        }

        @Override
        public void handleMessage(Message msg) {
            if (mp.getCurrentPosition() > pl.getCurrentPlayable().getMetadata().getEnd() &&
                    mp.getCurrentPosition() < pl.getCurrentPlayable().getMetadata().getEnd() + 1000) {
                psb.switchNext();
            } else {
                for (Playable.ProgressListener progressListener : NoteStream.PROGRESS_LISTENERS) {
                    progressListener.onProgressChanged(mp.getCurrentPosition());
                }
            }
        }
    }

    protected PhoneStateListener psl = new PhoneStateListener() {
        @Override
        public void onCallStateChanged(int state, String incomingNumber) {
            if (state == TelephonyManager.CALL_STATE_RINGING) {
                psb.pause();
            } else if(state == TelephonyManager.CALL_STATE_IDLE) {
                psb.play();
            } else if(state == TelephonyManager.CALL_STATE_OFFHOOK) {
                psb.pause();
            }
            super.onCallStateChanged(state, incomingNumber);
        }
    };

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            if (intent.getAction() == null) intent.setAction(Constants.ACTION_INIT);

            switch (intent.getAction()) {
                case Constants.ACTION_PLAY:
                    psb.play();
                    break;
                case Constants.ACTION_PAUSE:
                    psb.pause();
                    break;
                case Constants.ACTION_TOGGLE_PLAY:
                    psb.togglePlay();
                    break;
                case Constants.ACTION_STOP:
                    if (notification != null) {
                        notification.cancelNotification();
                    }
                    psb.stop();
                    break;
                case Constants.ACTION_SWITCH_PREVIOUS:
                    psb.switchPrevious();
                    break;
                case Constants.ACTION_SWITCH_NEXT:
                    psb.switchNext();
                    break;
                case Constants.ACTION_TOGGLE_REPEAT:
                    psb.toggleRepeat();
                    break;
                case Constants.ACTION_TOGGLE_SHUFFLE:
                    psb.setShuffle(!psb.doShuffle());
                    break;
                case Constants.ACTION_INIT:
                default:
                    mp.setWakeMode(this, PowerManager.PARTIAL_WAKE_LOCK);

                    mp.setOnCompletionListener(this);
                    progressHandler = new ProgressHandler(mp);

                    TelephonyManager tm = (TelephonyManager) getSystemService(TELEPHONY_SERVICE);
                    if(tm != null) {
                        tm.listen(psl, PhoneStateListener.LISTEN_CALL_STATE);
                    }

                    ms = new MediaSession(PlayerService.this, Constants.MEDIA_SESSION_TAG);
                    ms.setCallback(new NSMediaSessionCallback());
                    ms.setFlags(MediaSession.FLAG_HANDLES_MEDIA_BUTTONS | MediaSession.FLAG_HANDLES_TRANSPORT_CONTROLS);
                    ms.setActive(true);

                    progressTimer.scheduleAtFixedRate(new TimerTask() {
                        @Override
                        public void run() {
                            if (playing) {
                                progressHandler.obtainMessage(1).sendToTarget();
                            }
                        }
                    }, 1000, 1000);

                    notification = new NoteStreamNotification(this, this);
            }
        }

        return Service.START_STICKY;
    }

    @Override
    public void onDestroy() {
        TelephonyManager tm = (TelephonyManager) getSystemService(TELEPHONY_SERVICE);
        if(tm != null) {
            tm.listen(psl, PhoneStateListener.LISTEN_NONE);
        }

        mp.release();
        ms.release();
        progressTimer.cancel();
        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return psb = new PlayerServiceBinder();
    }

    public class NSMediaSessionCallback extends MediaSession.Callback {

        @Override
        public boolean onMediaButtonEvent(@NonNull Intent mediaButtonIntent) {
            String intentAction = mediaButtonIntent.getAction();
            if (Intent.ACTION_MEDIA_BUTTON.equals(intentAction)) {
                final KeyEvent event = mediaButtonIntent.getParcelableExtra(Intent.EXTRA_KEY_EVENT);
                if (event != null) {
                    if (event.getAction() == KeyEvent.ACTION_UP) {
                        // RELEASE

                        switch (event.getKeyCode()) {
                            case KeyEvent.KEYCODE_MEDIA_FAST_FORWARD:
                            case KeyEvent.KEYCODE_MEDIA_STEP_FORWARD:
                                psb.setProgress(Math.min(psb.getProgress() + PlayerService.DEFAULT_PROGRESS_CHANGE,
                                        psb.getCurrentPlayable().getMetadata().getLength()));
                                break;
                            case KeyEvent.KEYCODE_MEDIA_STEP_BACKWARD:
                                psb.setProgress(Math.max(psb.getProgress() - PlayerService.DEFAULT_PROGRESS_CHANGE, 0));
                                break;
                            case KeyEvent.KEYCODE_MEDIA_STOP:
                                psb.stop();
                                break;
                            case KeyEvent.KEYCODE_MEDIA_PLAY:
                                psb.play();
                                break;
                            case KeyEvent.KEYCODE_HEADSETHOOK:
                            case KeyEvent.KEYCODE_SPACE:
                            case KeyEvent.KEYCODE_MEDIA_PAUSE:
                                psb.pause();
                                break;
                            case KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE:
                                if (playing) {
                                    psb.pause();
                                } else {
                                    psb.play();
                                }
                                break;
                            case KeyEvent.KEYCODE_MEDIA_SKIP_FORWARD:
                            case KeyEvent.KEYCODE_MEDIA_NEXT:
                                psb.switchNext();
                                break;
                            case KeyEvent.KEYCODE_MEDIA_SKIP_BACKWARD:
                            case KeyEvent.KEYCODE_MEDIA_PREVIOUS:
                                psb.switchPrevious();
                                break;
                        }
                    }
                }
            }
            return super.onMediaButtonEvent(mediaButtonIntent);
        }
    }

    public class PlayerServiceBinder extends Binder {

        public Playlist shufflePlay(Playlist playlist) {
            Playlist old = pl;
            if (playlist != null && !playlist.isEmpty()) {
                mp.stop();
                pl.clear();
                pl.add(playlist);
                pl.skipTo(old.getCurrentPlayable());
                pl.setShuffle(true);
                pl.getCurrentPlayable().prepare(mp);

                for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(getCurrentPlayable());
                }
                play();
            }
            return old;
        }

        public Playlist playPlaylist(Playlist playlist) {
            Playlist old = pl;
            if (playlist != null && !playlist.isEmpty()) {
                pl.clear();
                pl.add(playlist);
                pl.skipTo(playlist.getCurrentPlayable());
                pl.getCurrentPlayable().prepare(mp);

                for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(getCurrentPlayable());
                }
                play();
            }
            return old;
        }

        public Playlist playAt(Playlist playlist, Playable playable) {
            Playlist old = pl;
            if (playlist != null) {
                pl.clear();
                pl.add(playlist);
                pl.skipTo(playable);
                pl.getCurrentPlayable().prepare(mp);

                for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(getCurrentPlayable());
                }
                play();
            }
            return old;
        }

        public Playlist play(final Playable playable) {
            return playPlaylist(Playlist.get(Playlist.TEMPORARY_PREFIX + playable.getID(), Collections.singletonList(playable)));
        }

        public void playNext(Playable playable) {
            pl.addNext(playable);
        }

        public void addToQueue(Playable playable) {
            pl.addRelative(++queueSize, playable);
        }

        public boolean play() {
            boolean old = playing;
            if (!pl.isEmpty()) {
                playing = true;
                mp.start();

                NoteStream.getInstance().library.lastListened.add(getCurrentPlayable());

                for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayStatusChanged(true);
                }
                if (!NoteStream.CONTROL_LISTENERS.contains(notification)) NoteStream.registerControlListener(notification);
            }
            return old;
        }

        public boolean togglePlay() {
            if (playing) {
                return pause();
            } else {
                return play();
            }
        }

        public boolean pause() {
            boolean old = !playing;
            playing = false;
            mp.pause();
            for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayStatusChanged(false);
            }
            return old;
        }

        public void stop() {
            for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayStatusChanged(false);
            }
            playing = false;
            mp.stop();
        }

        public Playlist setPlaylist(Playlist playlist) {
            Playlist old = pl;
            if (playlist != null && !playlist.isEmpty()) {
                Playable currentPlayable = null;
                if (!pl.isEmpty()) {
                    currentPlayable = pl.getCurrentPlayable();
                }
                pl.clear();
                pl.add(playlist);
                if (currentPlayable != null) {
                    pl.skipTo(currentPlayable);
                } else {
                    pl.skipTo(playlist.getCurrentPlayable());
                    pl.getCurrentPlayable().prepare(mp);
                    for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                        controlListener.onPlayableChanged(getCurrentPlayable());
                    }
                }
            }
            return old;
        }

        public boolean switchPrevious() {
            if (mp.getCurrentPosition() > 3000) {
                mp.seekTo(0);
                return false;
            } else {
                if (queueSize > 0) queueSize++;
                if (pl == null || pl.isEmpty()) return false;
                pl.switchPrevious().prepare(mp);
                if(isPlaying()) play();
                for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(getCurrentPlayable());
                }
                return true;
            }
        }

        public boolean switchNext() {
            if (queueSize > 0) queueSize--;
            if (pl == null || pl.isEmpty()) return false;
            pl.switchNext().prepare(mp);
            if(isPlaying()) play();
            for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayableChanged(getCurrentPlayable());
            }
            return true;
        }

        public void toggleRepeat() {
            repeatState = repeatState.next();
            switch (repeatState) {
                case NONE:
                case ALL:
                    mp.setLooping(false);
                    break;
                case ONE:
                    mp.setLooping(true);
                    break;
            }

            for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                controlListener.onRepeatStateChanged(repeatState);
            }
        }

        public RepeatState getRepeatState() {
            return repeatState;
        }

        public boolean setShuffle(boolean shuffle) {
            if (!pl.isEmpty()) {
                boolean old = pl.doShuffle();
                if (old != shuffle) queueSize = 0;
                pl.setShuffle(shuffle);
                for (Playable.ControlListener controlListener : NoteStream.CONTROL_LISTENERS) {
                    controlListener.onShuffleStateChanged(shuffle);
                }
                return old;
            }
            return false;
        }

        public boolean doShuffle() {
            return pl.doShuffle();
        }


        public Playable getCurrentPlayable() {
            if (!pl.isEmpty()) {
                return pl.getCurrentPlayable();
            }
            return null;
        }

        public void setProgress(int progress) {
            getCurrentPlayable().skipTo(mp, progress);
        }

        public int getProgress() {
            return mp.getCurrentPosition();
        }

        public boolean isEmpty() {
            return pl == null || pl.isEmpty();
        }

        public boolean isPlaying() {
            return playing;
        }
    }
}

File: app/src/main/java/hr/caellian/notestream/data/PlayerService.kt
/*
 * The MIT License (MIT)
 * NoteStream, android music player and streamer
 * Copyright (c) 2018 Tin Švagelj <tin.svagelj.email@gmail.com> a.k.a. Caellian
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

package hr.caellian.notestream.data

import android.app.Service
import android.content.Context
import android.content.Intent
import android.media.MediaPlayer
import android.media.session.MediaSession
import android.os.*
import android.support.annotation.RequiresApi
import android.telephony.PhoneStateListener
import android.telephony.TelephonyManager
import android.view.KeyEvent
import hr.caellian.notestream.NoteStream
import hr.caellian.notestream.data.playable.Playable
import hr.caellian.notestream.data.playlist.Playlist
import hr.caellian.notestream.data.playlist.PlaylistIterator
import hr.caellian.notestream.gui.Notification
import hr.caellian.notestream.lib.Constants
import hr.caellian.notestream.util.RepeatState
import java.util.*

class PlayerService : Service(), MediaPlayer.OnCompletionListener {
    internal var queueSize = 0

    private var ms: MediaSession? = null

    internal var playing: Boolean = false

    internal var notification: Notification? = null

    internal var progressHandler: Handler? = null

    private var psl: PhoneStateListener = object : PhoneStateListener() {
        var previousState = false

        override fun onCallStateChanged(state: Int, incomingNumber: String) {
            when (state) {
                TelephonyManager.CALL_STATE_IDLE -> if (previousState) psb?.play()
                else -> previousState = psb?.pause()?.not() ?: false
            }
        }
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        if (repeatState != RepeatState.ONE) {
            if (iterator?.hasNext() == false && repeatState != RepeatState.ALL) {
                psb!!.pause()
            } else {
                psb!!.switchNext()
            }
        }
    }

    private class ProgressHandler(internal var mp: MediaPlayer) : Handler() {

        override fun handleMessage(msg: Message) {
            if (mp.currentPosition > iterator?.current()?.info?.end ?: iterator?.current()?.info?.length ?: 0 && mp.currentPosition < iterator?.current()?.info?.end ?: 0 + 1000) {
                psb!!.switchNext()
            } else {
                for (progressListener in NoteStream.PROGRESS_LISTENERS) {
                    progressListener.onProgressChanged(mp.currentPosition)
                }
            }
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent != null) {
            if (intent.action == null) intent.action = Constants.ACTION_INIT

            when (intent.action) {
                Constants.ACTION_PLAY -> psb!!.play()
                Constants.ACTION_PAUSE -> psb!!.pause()
                Constants.ACTION_TOGGLE_PLAY -> psb!!.togglePlay()
                Constants.ACTION_STOP -> {
                    if (notification != null) {
                        notification!!.cancelNotification()
                    }
                    psb!!.stop()
                }
                Constants.ACTION_SWITCH_PREVIOUS -> psb!!.switchPrevious()
                Constants.ACTION_SWITCH_NEXT -> psb!!.switchNext()
                Constants.ACTION_TOGGLE_REPEAT -> psb!!.toggleRepeat()
                Constants.ACTION_TOGGLE_SHUFFLE -> psb!!.setShuffle(!psb!!.doShuffle())
                Constants.ACTION_INIT -> {
                    mp.setWakeMode(this, PowerManager.PARTIAL_WAKE_LOCK)

                    mp.setOnCompletionListener(this)
                    progressHandler = ProgressHandler(mp)

                    val tm = getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager?
                    tm?.listen(psl, PhoneStateListener.LISTEN_CALL_STATE)

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        ms = MediaSession(this, Constants.MEDIA_SESSION_TAG)
                        ms?.setCallback(NSMediaSessionCallback())
                        ms?.isActive = true
                    }

                    progressTimer.scheduleAtFixedRate(object : TimerTask() {
                        override fun run() {
                            if (playing) {
                                progressHandler?.obtainMessage(1)?.sendToTarget()
                            }
                        }
                    }, 1000, 1000)

                    notification = Notification(this, this)
                }
                else -> {
                    mp.setWakeMode(this, PowerManager.PARTIAL_WAKE_LOCK)
                    mp.setOnCompletionListener(this)
                    progressHandler = ProgressHandler(mp)
                    val tm = getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager?
                    tm?.listen(psl, PhoneStateListener.LISTEN_CALL_STATE)

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        ms = MediaSession(this@PlayerService, Constants.MEDIA_SESSION_TAG)
                        ms?.setCallback(NSMediaSessionCallback())
                        ms?.isActive = true
                    }

                    progressTimer.scheduleAtFixedRate(object : TimerTask() {
                        override fun run() {
                            if (playing) {
                                progressHandler?.obtainMessage(1)?.sendToTarget()
                            }
                        }
                    }, 1000, 1000)
                    notification = Notification(this, this)
                }
            }
        }

        return Service.START_STICKY
    }

    override fun onDestroy() {
        val tm = getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager?
        tm?.listen(psl, PhoneStateListener.LISTEN_NONE)

        mp.release()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            ms?.release()
        }
        progressTimer.cancel()
        super.onDestroy()
    }

    override fun onBind(intent: Intent): IBinder? {
        psb = PlayerServiceBinder()
        return psb
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    inner class NSMediaSessionCallback : MediaSession.Callback() {

        override fun onMediaButtonEvent(mediaButtonIntent: Intent): Boolean {
            val intentAction = mediaButtonIntent.action
            if (Intent.ACTION_MEDIA_BUTTON == intentAction) {
                val event = mediaButtonIntent.getParcelableExtra<KeyEvent>(Intent.EXTRA_KEY_EVENT)
                if (event != null) {
                    if (event.action == KeyEvent.ACTION_UP) {
                        // RELEASE

                        when (event.keyCode) {
                            KeyEvent.KEYCODE_MEDIA_FAST_FORWARD, KeyEvent.KEYCODE_MEDIA_STEP_FORWARD -> psb!!.progress = Math.min(psb!!.progress + PlayerService.DEFAULT_PROGRESS_CHANGE,
                                    psb!!.currentPlayable!!.info.length)
                            KeyEvent.KEYCODE_MEDIA_STEP_BACKWARD -> psb!!.progress = Math.max(psb!!.progress - PlayerService.DEFAULT_PROGRESS_CHANGE, 0)
                            KeyEvent.KEYCODE_MEDIA_STOP -> psb!!.stop()
                            KeyEvent.KEYCODE_MEDIA_PLAY -> psb!!.play()
                            KeyEvent.KEYCODE_HEADSETHOOK, KeyEvent.KEYCODE_SPACE, KeyEvent.KEYCODE_MEDIA_PAUSE -> psb!!.pause()
                            KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE -> if (playing) {
                                psb!!.pause()
                            } else {
                                psb!!.play()
                            }
                            KeyEvent.KEYCODE_MEDIA_SKIP_FORWARD, KeyEvent.KEYCODE_MEDIA_NEXT -> psb!!.switchNext()
                            KeyEvent.KEYCODE_MEDIA_SKIP_BACKWARD, KeyEvent.KEYCODE_MEDIA_PREVIOUS -> psb!!.switchPrevious()
                        }
                    }
                }
            }
            return super.onMediaButtonEvent(mediaButtonIntent)
        }
    }

    inner class PlayerServiceBinder : Binder() {
        var repeatState: RepeatState = RepeatState.NONE

        val currentPlayable: Playable?
            get() = iterator?.current()

        var progress: Int
            get() = mp.currentPosition
            set(progress) {
                currentPlayable?.skipTo(mp, progress)
            }

        val isEmpty: Boolean
            get() = iterator?.isEmpty ?: true

        val isPlaying: Boolean
            get() = playing

        fun shufflePlay(playlist: Playlist?): Playlist? {
            val old = iterator?.playlist
            if (playlist != null && !playlist.isEmpty) {
                mp.stop()
                iterator = PlaylistIterator.Random(playlist)
                iterator?.current()?.prepare(mp)

                for (controlListener in NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(currentPlayable)
                }
                play()
            }
            return old
        }

        fun playPlaylist(playlist: Playlist?): Playlist? {
            val old = iterator?.playlist
            if (playlist != null && !playlist.isEmpty) {
                iterator?.reassign(playlist)
                iterator?.current()?.prepare(mp)

                for (controlListener in NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(currentPlayable)
                }
                play()
            }
            return old
        }

        fun playAt(plIterator: PlaylistIterator, playable: Playable): Playlist? {
            val old = iterator?.playlist
            iterator = plIterator
            iterator?.switchTo(playable)?.current()?.prepare(mp)

            for (controlListener in NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayableChanged(currentPlayable)
            }
            play()
            return old
        }

        fun playAt(playlist: Playlist?, playable: Playable?): Playlist? {
            val old = iterator?.playlist
            if (playlist != null) {
                iterator?.reassign(playlist)
                iterator?.switchTo(playable)?.current()?.prepare(mp)

                for (controlListener in NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(currentPlayable)
                }
                play()
            }
            return old
        }

        fun play(playable: Playable): Playlist? {
            return playPlaylist(Playlist.get(Constants.PLAYLIST_TEMPORARY_PREFIX + playable.id, data = listOf(playable)))
        }

        fun playNext(playable: Playable) {
            iterator?.add(playable, iterator!!.current + 1)
        }

        fun playNext(playlist: Playlist) {
            iterator?.addAll(playlist.playlist, iterator!!.current + 1)
        }

        fun addToQueue(playable: Playable) {
//            iterator?.queue(playable, iterator.current + ++queueSize)
        }

        fun play(): Boolean {
            val old = playing
            if (iterator?.isEmpty == false) {
                playing = true
                mp.start()

                currentPlayable?.let {
                    NoteStream.instance.data.lastListened.add(it)
                }

                for (controlListener in NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayStatusChanged(true)
                }

                notification?.let {
                    NoteStream.registerControlListener(it)
                }
            }
            return old
        }

        fun togglePlay(): Boolean {
            return if (playing) {
                pause()
            } else {
                play()
            }
        }

        fun pause(): Boolean {
            val old = !playing
            playing = false
            mp.pause()
            for (controlListener in NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayStatusChanged(false)
            }
            return old
        }

        fun stop() {
            for (controlListener in NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayStatusChanged(false)
            }
            playing = false
            mp.stop()
        }

        fun setPlaylist(playlist: Playlist?): Playlist? {
            val old = iterator?.playlist
            if (playlist != null && !playlist.isEmpty) {
                iterator?.reassign(playlist)
            }
            return old
        }

        fun setIterator(playlistIterator: PlaylistIterator): PlaylistIterator? {
            val old = iterator
            iterator = playlistIterator
            return old
        }

        fun switchPrevious(): Boolean {
            if (mp.currentPosition > 3000) {
                mp.seekTo(0)
                return false
            } else {
                if (queueSize > 0) queueSize++
                if (iterator == null || iterator?.isEmpty == true) return false
                iterator?.switchPrevious()?.prepare(mp)
                if (isPlaying) play()
                for (controlListener in NoteStream.CONTROL_LISTENERS) {
                    controlListener.onPlayableChanged(currentPlayable)
                }
                return true
            }
        }

        fun switchNext(): Boolean {
            if (queueSize > 0) queueSize--
            if (iterator == null || iterator?.isEmpty == true) return false
            iterator?.switchNext()?.prepare(mp)
            if (isPlaying) play()
            for (controlListener in NoteStream.CONTROL_LISTENERS) {
                controlListener.onPlayableChanged(currentPlayable)
            }
            return true
        }

        fun toggleRepeat() {
            repeatState = repeatState.next()
            when (repeatState) {
                RepeatState.NONE, RepeatState.ALL -> mp.isLooping = false
                RepeatState.ONE -> mp.isLooping = true
            }

            for (controlListener in NoteStream.CONTROL_LISTENERS) {
                controlListener.onRepeatStateChanged(repeatState)
            }
        }

        fun setShuffle(shuffle: Boolean): Boolean {
            if (iterator?.isEmpty == false) {
                val old = iterator is PlaylistIterator.Random
                if (old != shuffle) queueSize = 0
                iterator = iterator?.setRandom(shuffle)
                for (controlListener in NoteStream.CONTROL_LISTENERS) {
                    controlListener.onShuffleStateChanged(shuffle)
                }
                return old
            }
            return false
        }

        fun doShuffle(): Boolean {
            return iterator?.id == Constants.ITERATOR_RANDOM_ID
        }
    }

    companion object {
        // http://www.tutorialsface.com/2015/08/android-custom-notification-tutorial/
        const val DEFAULT_PROGRESS_CHANGE = 10000

        private val progressTimer = Timer()

        var iterator: PlaylistIterator? = null

        internal var mp = MediaPlayer()
        internal var repeatState = RepeatState.NONE

        internal var psb: PlayerServiceBinder? = null
    }
}
